### DISCUSSION TITLE
How to Debug for Mac Desktop

### DISCUSSION THREAD

**Discussion by Bilbosax** [airsdk/Adobe-Runtime-Support#756] [General]
I am writing my first AIR desktop app for Mac. It runs great on Windows, so I have now transferred the project over to Mac to try and get it released on that platform. I understand that to get it distributed, I have to go through the painful process of notarizing it for Catalina because I will not be using the App Store. But my problem at the moment is that I can't even debug or run the program on Mac. I have installed flex, overlayed the latest AIR SDK by copy/paste/merge, transferred my project in IntelliJ from Windows. But if I click on the "Run" or "Debug" buttons, it fails every time.  First, it makes an isssue that ADL is not safe and might be malicious.  I got into security settings to allow it.  I get the same error for the framework, and allow it also.  Finally, after showing the Harman logo, it fails stating that it can't find several components like StyleableTextField and LabelItemRenderer. I am wondering, to simply test and debug on Mac, do you have to have some kind of signature in IntelliJ beyond the self-signed signature checkbox?? Or might I have the SDK improperly installed for Mac?? Could really use some help because AIR on Mac is brand new to me, so I need help sidestepping the potholes.

VerifyError: Error #1014: Class spark.components.supportClasses::StyleableTextField could not be found.

at flash.display::MovieClip/nextFrame()

at mx.managers::SystemManager/deferredNextFrame()[D:\flex_sdk\Sources\flex-sdk\frameworks\projects\framework\src\mx\managers\SystemManager.as:286]

at mx.managers::SystemManager/preloader_preloaderDocFrameReadyHandler()[D:\flex_sdk\Sources\flex-sdk\frameworks\projects\framework\src\mx\managers\SystemManager.as:2635]

at flash.events::EventDispatcher/dispatchEventFunction()

at flash.events::EventDispatcher/dispatchEvent()

at mx.preloaders:😛reloader/timerHandler()[D:\flex_sdk\Sources\flex-sdk\frameworks\projects\framework\src\mx\preloaders\Preloader.as:525]

at flash.utils::Timer/_timerDispatch()

at flash.utils::Timer/tick()

**Comment by Bilbosax** [comment]
A little more information.  I was able to create an empty project that successfully compiled, so I think I have the SDK set up correctly. So I tried copying all of the windows mxml and actionscript files to the new empty project, and I still get all the same errors, can't find StyleableTextField, etc. What I find interesting is that the addresses where it appears to look for files in the error are Windows addresses, not Mac addresses:

"D:\flex_sdk\Sources\flex-sdk\frameworks\projects\framework\src\mx\managers\SystemManager.as:286"

This file does not exist on my Mac Machine of course. Is this the EXACT address that it is looking for things, or just an error reference address? If it is an exact address, why would a Flex for Mac SDK and and AIR for Mac overlay be pointing to Windows addresses on a Mac??

**Comment by Bilbosax** [comment]
OK, I forgot to add a dependency for a swc, so my initial problem is solved, but now I have a serious issue and absolutely no errors being thrown to figure out a problem.

My program now compiles without error, the Harman splash screen displays momentarily and then disappears.... and that is it. No app. My program doesn't show at all. ADL is sitting down in the dock running, but there is no application displaying on screen or displayed in the dock. Ever seen anything like this???

In run mode, it shows Distriqt's webview ANE initializing which looks successful, and then nothing.  In debug mode, it just says "Player connected; session starting."  No errors, but also no app.

I am a rookie on Mac.  I don't know if anything needs to be set differently in the descriptor for Mac.  InitialWindow settings all look fairly straight forward to me.  Could really use some help here @ajwfrost 

I'll take wild guesses ;)

**Comment by Bilbosax** [comment]
So, I now have a compiling and visible app. It was not the fault of an ANE, but a visibility issue I am currently sorting out.

@ajwfrost One question you could answer for me.  I maximize the app upon opening it using NativeWindow.maximize(). In the development environment, it does not create its own "desktop" on a Mac as it usually does when you maximize an app, and the bottom extends behind the dock. Will this also happen when I deploy the app to users, or is it only happening because I am in a development environment and debugging inside IntelliJ/ADL?

**Reply by salbeik** [reply]
I will just let you have more time .. .the last 3 times you just solved it yourself :D
Will this also happen when I deploy the app to users
-There exists methods to discover the hight of the dock ... I think also for Mac. I can tell you this much, I have a app that I just mazimized and it didn't lay behind the dock if the dock is set to be there all the time. If I put my dock to hide automatically it will lay behind the dock . But its a very simple thing to test yourself, simply pack the app and test it.

**Comment by Bilbosax** [comment]
Thanks for the information @salbeik.  It would be easy to test if I had a clue of how to notarize the app and package it to deploy on a mac, but unfortunately I don't know how at this time.  I haven't gotten that far.

**Reply by salbeik** [reply]
well you can just build the app in inteliji, it will react the same as if notarized. 

I can help with the process though (but if there's something to consider about ANE I dont know that part), what do you need ? for notarizing I simply do this for my app called NotesandMath:

you must have Xcode 3.app installed - call it like i do:
sudo xcode-select -s /Volumes/Disk\ 2TB/Xcode\ 3.app/

xcrun altool --notarize-app --primary-bundle-id "com.notesandmath.application.zip" --username "your-email-username" --password "@keychain:AC_PASSWORD" --file NotesandMath.zip

Unzip notorized zip file, and staple it:

xcrun stapler staple Notes\ and\ Math.app/

Verify everything:
codesign -dv --verbose=4 Notes\ and\ Math.app/

**Comment by Bilbosax** [comment]
@salbeik , again thanks for your help.  This will be a tremendous resource for me as I am a Mac newbie.  A Couple of quick questions:

1)  I have XCode installed on my Mac, but not Xcode 3.app.  Where do you get that?
2) I don't want to use the Mac App Store but want to deliver the app directly from my website. To notarize for Mac without going through the App Store, do I need to get both a Developer ID Installer and Developer ID Application certificate.  Do you know how to handle these?

I am developing in IntelliJ if that helps at all.

**Reply by salbeik** [reply]
I also write my apps in IntelliJ, so I do all these things to be able to publish my app :
++++++++++++++++++
Open MacStore > search for xcode > download

yeah the whole code signing, was for me extremely hard to figure out. If you dont write your code in Xcode, its not easy. So this is what I do:
++++++++++++++++
PACKAGE NATIVE - caoptive runtime

Location: stand inside the src folder of NotesMath source folder:
/Notes_Math/src/

adt -package -keystore certificate.p12 -storetype pkcs12 -target bundle Notes\ and\ Math.app Notes_Math-app.xml .

I have attached my icons.icon,entitlements.plist and Info.plist so that you can see how they must look - otherwise you get a error.

[Nyt arkiv.zip](https://github.com/Gamua/Adobe-Runtime-Support/files/6252008/Nyt.arkiv.zip)

If you have updated the version number you need to edit the Info.plist file and add the new version number.

cp icons.icns Notes\ and\ Math.app/Contents/Resources/.   (So I have a copy of the  icons.icns file, where I have added extra high resolution icons that are  nessecary)

cp Info.plist Notes\ and\ Math.app/Contents/.  

rm Notes\ and\ Math.app/Contents/Frameworks/Adobe\ AIR.framework/Versions/Current/Resources/WebKit.dylib

chmod -R 777 Notes\ and\ Math.app/

cp Info_flash.plist Notes\ and\ Math.app/Contents/Frameworks/Adobe\ AIR.framework/Versions/1.0/Resources/Flash\ Player.plugin/Contents/Info.plist

codesign -f -v -s 'Developer ID Application: Celsius IT ApS (xxx)' --timestamp Notes\ and\ 
Math.app/Contents/Frameworks/Adobe\ AIR.framework/Versions/Current/Resources/WebKit.dylib

codesign -f -v -s 'Developer ID Application: Celsius IT ApS (xxx)' --timestamp Notes\ and\ Math.app/Contents/Frameworks/Adobe\ AIR.framework/Versions/1.0/Resources/AdobeCP15.plugin

codesign -f -v -s 'Developer ID Application: Celsius IT ApS (xxx)' --timestamp Notes\ and\ Math.app/Contents/Frameworks/Adobe\ AIR.framework/Versions/1.0/Resources/Flash\ Player.plugin/

codesign -f -v -s 'Developer ID Application: Celsius IT ApS (xxx)' --timestamp Notes\ and\ Math.app/Contents/Frameworks/Adobe\ AIR.framework/Versions/1.0/Resources/adobecp.plugin

codesign -f -v -s 'Developer ID Application: Celsius IT ApS (xxx)' --timestamp Notes\ and\ Math.app/Contents/Frameworks/Adobe\ AIR.framework/Versions/1.0/Adobe\ AIR_64 

codesign -f -v -s 'Developer ID Application: Celsius IT ApS (xxx)' --timestamp Notes\ and\ Math.app/Contents/Frameworks/Adobe\ AIR.framework/Versions/1.0

codesign -f -v -s 'Developer ID Application: Celsius IT ApS (xxx)' --timestamp Notes\ and\ Math.app/Contents/Frameworks/Adobe\ AIR.framework/Versions/1.0/Resources/A2712Enabler

codesign --force  --verbose=9 -s "Developer ID Application: Celsius IT ApS (xxx)" --options runtime --deep Notes\ and\ Math.app/Contents/Frameworks/Adobe\ AIR.framework/Versions/1.0/Resources/A2712Enabler

codesign -f -v -s 'Developer ID Application: Celsius IT ApS (xxx)' --timestamp Notes\ and\ Math.app

codesign --force  --verbose=9 -s "Developer ID Application: Celsius IT ApS (xxx)" --options runtime --deep --entitlements entitlements.plist --timestamp Notes\ and\ Math.app/Contents/MacOS/Notes\ and\ Math

codesign --force  --verbose=9 -s "Developer ID Application: Celsius IT ApS (xxx)" --options runtime --deep --entitlements entitlements.plist --timestamp Notes\ and\ Math.app

codesign -f -v -s '3rd Party Mac Developer Application: Celsius IT ApS (xxx)' --timestamp --entitlements entitlements.plist Notes\ and\ Math.app/

codesign --display --entitlements - Notes\ and\ Math.app

codesign -dvv Notes\ and\ Math.app


And then i  notarize. I put all these commands above in script that I just run.
https://developer.apple.com/documentation/xcode/notarizing_macos_software_before_distribution#3087722

**Reply by salbeik** [reply]
You get the certificates here: https://developer.apple.com > press account and sign in.
